includes/admin_import.php

```diff
@@ -164,31 +165,61 @@ function tsml_import_page()
             $data_source_parent_region_id = (int) $_POST['tsml_add_data_source_parent_region_id'];
             $data_source_change_detect = sanitize_text_field($_POST['tsml_add_data_source_change_detect']);
 
+            //initialize variables 
+            $import_meetings = $delete_meeting_ids = [];
+            
             //try fetching	
             $response = wp_safe_remote_get($data_source_url, [
                 'timeout' => 30,
                 'sslverify' => false,
             ]);
-            if (is_array($response) && !empty($response['body']) && ($body = json_decode($response['body'], true))) {
+            if (is_array($response) && !empty($response['body']) && ($meetings = json_decode($response['body'], true))) {
 
-                //if already set, hard refresh
-                if (array_key_exists($data_source_url, $tsml_data_sources)) {
-                    tsml_delete(tsml_get_data_source_ids($data_source_url));
+                //allow reformatting as necessary
+                $meetings = tsml_sanitize_import_meetings($meetings, $data_source_url, $data_source_parent_region_id);
+
+                //actual meetings to import
+                $import_meetings = array();
+                $updated_meeting_count = 0;
+
+                //for new data sources we import every meeting, otherwise test for changed meetings
+                if (!array_key_exists($data_source_url, $tsml_data_sources)) {
+                    $import_meetings = $meetings;
+                } else {
+                    
+                    //get updated feed import record set 
+                    list($import_meetings, $delete_meeting_ids, $change_log) = tsml_get_changed_import_meetings($meetings, $data_source_url, $data_source_parent_region_id);
+
+                    //drop meetings that weren't found in the import
+                    if (count($delete_meeting_ids)) {
+                        tsml_delete($delete_meeting_ids);
+                    }
+                    
                     tsml_delete_orphans();
+
+                    if (count($change_log) && $tsml_debug) {
+                        $message = tsml_build_import_change_report($data_source_name, $change_log);
+                        tsml_alert($message, 'info');
+                    }
+                    
+                    //empty change log means we're up to date
+                    if (!count($change_log)) {
+                        tsml_alert(__('Your meeting list is already in sync with the feed.', '12-step-meeting-list'), 'success');                    
+                    }
                 }
 
                 $tsml_data_sources[$data_source_url] = [
                     'status' => 'OK',
                     'last_import' => current_time('timestamp'),
-                    'count_meetings' => 0,
+                    'count_meetings' => count($import_meetings),
```

Comment:
nbd but i wonder if this is too hasty now to be setting  the finished number of meetings? should we do this instead after the real number has been modified, in case something goes wrong?

Reply:
Yeah, I changed this yesterday when I saw the number wasn't updating in the Ajax step. I can move it there, it'll just have to query all source meetings for an accurate count after each AJAX import run.